components/brave_wallet/browser/zcash/zcash_keyring.cc

```diff
@@ -38,6 +43,101 @@ std::optional<std::vector<uint8_t>> ZCashKeyring::GetPubkey(
   return hd_key_base->GetPublicKeyBytes();
 }
 
+std::optional<std::vector<uint8_t>> ZCashKeyring::GetPubkeyHash(
+    const mojom::ZCashKeyId& key_id) {
+  auto hd_key_base = DeriveKey(key_id);
+  if (!hd_key_base) {
+    return std::nullopt;
+  }
+
+  return Hash160(hd_key_base->GetPublicKeyBytes());
+}
+
+#if BUILDFLAG(ENABLE_ORCHARD)
+
+std::optional<std::string> ZCashKeyring::GetUnifiedAddress(
+    const mojom::ZCashKeyId& transparent_key_id,
+    const mojom::ZCashKeyId& zcash_key_id) {
+  if (!orchard_key_) {
+    NOTREACHED();
+    return std::nullopt;
+  }
+
+  auto esk = orchard_key_->DeriveHardenedChild(zcash_key_id.account);
+  if (!esk) {
+    return std::nullopt;
+  }
+
+  auto orchard_addr_bytes =
+      zcash_key_id.change ? esk->GetDiversifiedAddress(zcash_key_id.index,
+                                                       OrchardKind::Internal)
+                          : esk->GetDiversifiedAddress(zcash_key_id.index,
+                                                       OrchardKind::External);
+  if (!orchard_addr_bytes) {
+    return std::nullopt;
+  }
+
+  auto transparent_pubkey_hash = GetPubkeyHash(transparent_key_id);
+  if (!transparent_pubkey_hash) {
+    return std::nullopt;
+  }
+
+  return GetMergedUnifiedAddress(
+      std::vector<ParsedAddress>{
+          ParsedAddress(ZCashAddrType::kP2PKH, transparent_pubkey_hash.value()),
+          ParsedAddress(ZCashAddrType::kOrchard,
+                        std::vector<uint8_t>(orchard_addr_bytes->begin(),
+                                             orchard_addr_bytes->end()))},
+      testnet_);
+}
+
+mojom::ZCashAddressPtr ZCashKeyring::GetShieldedAddress(
+    const mojom::ZCashKeyId& key_id) {
+  if (!orchard_key_) {
+    NOTREACHED();
+    return nullptr;
+  }
+
+  auto esk = orchard_key_->DeriveHardenedChild(key_id.account);
```

Comment:
Is it something that is CPU time consuming? Should we cache it then?

Reply:
CPU consuming part i was talking about is zero proof generation. This one is just a key derivation which shouldn't be.

Reply:
This doesn't seem computationally expensive. It's essentially doing a Blake2b-512 call (a single hash). Here's the [implementation](https://github.com/zcash/orchard/blob/2b9c9a1deb66f8b20cd5c07fdd0cec87895f5c16/src/zip32.rs#L206-L233)

Here's the spec: https://zips.z.cash/zip-0032#orchard-child-key-derivation PRFexpand is essentially the blake2b call.